container/apbf/filter.go

```diff
@@ -0,0 +1,616 @@
+	// items.
+	f.mtx.Lock()
+	if f.itemsInCurGeneration == f.itemsPerGeneration {
+		f.nextGeneration()
+	}
+	f.itemsInCurGeneration++
+
+	// Set the relevant bits in the filter for 'k' slices, starting from the
+	// first slice (slice 0), as determined by the equivalent of a separate hash
+	// (index) function for each slice.  See the comments in deriveIndex for
+	// more details about the enhanced double hashing technique used.
+	//
+	// Note that this implementation uses a ring buffer where the position index
+	// of the first slice is modified to simulate shifting, so the code below
+	// uses logical slices to map to the correct physical location within the
+	// filter data.
+	//
+	// As a further optimization, rather than fully calculating each independent
+	// bit index for each slice via the closed formula for enhanced double
+	// hashing, this interleaves the calculation with the loop over the slices
+	// which benchmarking shows is about 30% faster.
+	logicalSlice := f.baseIndex
+	sliceBitOffset := uint64(logicalSlice) * f.bitsPerSlice
+	derivedIdx, acc := deriveIndex(logicalSlice, hash1, hash2)
+	for i := uint8(0); i < f.k; i++ {
+		f.setBit(sliceBitOffset + derivedIdx%f.bitsPerSlice)
+
+		// Move to the next logical slice while wrapping around the ring buffer
+		// if needed.
+		logicalSlice++
+		if logicalSlice == f.kPlusL {
+			logicalSlice = 0
+			sliceBitOffset = 0
+
+			// Reset the derived bit index using enhanced double hashing
+			// accordingly.
+			derivedIdx, acc = hash1, hash2
+			continue
+		}
+		sliceBitOffset += f.bitsPerSlice
+
+		// Derive the next bit index using enhanced double hashing.
+		derivedIdx += acc
+		acc += uint64(logicalSlice)
+	}
+	f.mtx.Unlock()
+}
+
+// isBitSet returns whether or not the bit at the provided absolute index in the
+// underlying filter data is set.
+//
+// This function MUST be called with the filter mutex held.
+func (f *Filter) isBitSet(bit uint64) bool {
+	return f.data[bit>>3]&(1<<(bit&7)) != 0
+}
+
+// Contains returns the result of a probabilistic membership test of the
+// provided data such that there is a non-zero probability of false positives,
+// per the false positive rate of the filter, and a zero probability of false
+// negatives for the previous number of items supported by the max capacity of
+// the filter.
+//
+// In other words, the most recent max capacity number of items added to the
+// filter will always return true while items that were never added or have been
+// expired will only report true with the false positive rate of the filter.
+//
+// This function is safe for concurrent access.
+func (f *Filter) Contains(data []byte) bool {
+	// Calculate the hash of the data for use below in selecting the bits to
+	// test.  Note that this is intentionally done prior to taking the filter
+	// lock to increase performance when there is contention.
+	hash1, hash2 := siphash.Hash128(f.key0, f.key1, data)
```

Comment:
Hah, I noticed this and fixed it as you were reviewing.  If you refresh, you should see the change.

Reply:
To expand on the previous comment for reference, there originally was no key, but I decided it really should have one to prevent malicious behavior and just hadn't updated these to account for it.

I decided to simply gate it by the single mutex since benchmarks showed that it made basically no discernible difference in practice, even with 16 cores hammering against it, as the difference fell withing the margin of error.